### DISCUSSION TITLE
MVC in a single file by "package::" in Mojolicious::Lite?

### DISCUSSION THREAD

**Discussion by westmj** [mojolicious/mojo#1689] [Q&A]
From https://docs.mojolicious.org/Mojolicious/Guides/Growing I want to perversely avoid modules (*.pm) by using packages in a single "Lite" script,  and still do testing.  I managed to get partway there as zipped below by isolating the Model of MVC into the package Model::Users in login.pl following the Growing example, and can test it with t/login.t with only minor changes.

cd login
chmod 755 login.pl
prove t/login.t
./login.pl daemon

(but "morbo login.pl" is a disaster...)

But when I tried to isolate the Controller::Login, leaving the View in package main in login_mvc.pl , I have something wrong... ouch. It is hard for me to understand the deflation, so to speak. Can anyone help? I hope your teeth don't hurt when you look at this. Thanks.  I feel like I am almost there.  I am trying to up my game...

chmod 755 login_mvc.pl
etc... trying the passwords from a browser...

Thanks from a newbie...

[login.zip](https://github.com/mojolicious/mojo/files/5929555/login.zip)

**Comment by kiwiroy** [comment]
In `login.pl` move the `use Mojolicious::Lite -signatures;` to within `main`. `Model::Users` needs `strict`, `warnings` and `signatures` [enabling](https://git.io/Jtggz).

If you really wish to have a single file full app in `login_mvc.pl`, the *app* needs to be a class inheriting from `Mojolicious` and the script `main` needs to be the *script* from the growing guide, returning the instance. There's a few other syntax changes similar to above [here](https://git.io/JtgaC).

**Reply by westmj** [reply]
Thanks!  And your version runs fine under "morbo login_mvc.pl", but my login.pl spews errors.  How come mine sucks under morbo but not as a daemon?

And then there are tests.  I made only a slight change in the test files, t/login[_mvc].t     The"prove t/login.t" passes, but your script with "prove t/login_mvc.pl" does not.  I am a perplexed newbie.

And could these single-file scripts be inflated into applications with separation into modules Model::Users and Login::Controller::Login ?  And why choose that over Controller::Login ? But not choose Login::Model::Users ? 

Thanks for any help.

**Reply by westmj** [reply]
okay, I am working on it... give me some more time...  :-)     Caution: Newbie at work...
